### ISSUE TITLE
Automatically reload certificate and private key when they are updated in the filesystem

### ISSUE THREAD

**Issue by briansmith** [runconduit/conduit#369]
This is a follow-up to #306. When a new Kubernetes secret with a new private key and/or certificate is distributed to the proxy (through the kubelet), we need to load the new private key and certificate.

Since the private key and certificate will be provided to the proxy through the filesystem (via a filesystem-mapped Kubernetes secret), we can use `inotify` to watch the given paths for changes. It is important to use `inotify` or an equivalent mechanism whenever possible because we don't want to have a polling background thread in normal use. However, in the case that the `inotify` subscription fails, or on platforms without `inotify`, we should have a backup implementation that does polling in a background thread or equivalent.

It will be instructive to study the implementation of similar measures within linkerd, in particular `inotify`-related bugs.

This builds upon #306; #306 needs to be done first.

**Comment by briansmith** [comment]
On the consuming side, we already have this:
https://github.com/runconduit/conduit/blob/b114ef6819e98a5f1840f50e58193e8d02716fc2/proxy/src/connection.rs#L128

The idea is that configuration objects are immutable and so we can clone (an `Arc` for) the current configuration and use it for connections totally independently from any configuration changes.

Then what's left is to load a new configuration when the configuration files change, and then to load the configuration file asynchronously, instead of at proxy startup.

https://github.com/runconduit/conduit/blob/b114ef6819e98a5f1840f50e58193e8d02716fc2/proxy/src/lib.rs#L235-L242

2. Have the control thread poll for changes to the existence of any of the files that would be loaded on disk. When we see any of them has changed, reload all of them from disk. When we get a valid configuration, as determined by `load_from_disk()`, then update the "current" configuration to the one returned by `load_from_disk()`.

@olix0r suggested that we simply have the control thread work on a timer and avoid `inotify` for now, and add `inotify`-based watching later. That sounds OK to me. The only issue is that, especially when we don't have a configuration loaded yet, we'd have to poll often to minimize the time we can't do TLS.

I suggest that the control thread and the I/O threads share the configuration using a `Arc<Mutex<Arc<Option<CommonConfig>>>>` or (probably simpler) equivalent, rather than passing it on a channel. The configuration almost never changes so a mutex wouldn't have any contention, so I don't see using a channel as having any performance or scalability advantage over the simpler thing.

**Comment by briansmith** [comment]
Another issue that came up is: What do we do when we currently have a valid configuration, but then we try to load an updated configuration, and the new configuration fails. That could happen, for example, if the certificate file is updated and the private key has not been updated *yet*; in that case we should just keep using the old configuration until the new valid configuration is ready. That could also happen if there was a problem with the old configuration and we're supposed to stop using it; in that case, we should stop using the old configuration.

For the short term, considering other factors, I think it is OK to just always use the old configuration unless/until we get an updated configuration that is actually valid.

**Comment by hawkw** [comment]
@briansmith 
> I suggest that the control thread and the I/O threads share the configuration using a Arc<Mutex<Arc<Option<CommonConfig>>>> or (probably simpler) equivalent, rather than passing it on a channel. The configuration almost never changes so a mutex wouldn't have any contention, so I don't see using a channel as having any performance or scalability advantage over the simpler thing.

I was thinking this seemed like a good use-case for [`futures-watch`](https://github.com/carllerche/better-future/blob/master/futures-watch/src/lib.rs), which seems like the simplest option from Conduit's perspective...
